slangpy/torchintegration/torchtensormarshall.py

```diff
@@ -1 +1,272 @@
+            interop_tensor.copy_(primal)
+
+            primal_calldata = {
+                "buffer": data.interop_buffer,
+                "layout": {"offset": 0, "strides": strides},
+                "_shape": shape,
+            }
+
+            if not self.d_in and not self.d_out:
+                return primal_calldata
+
+            result = {"primal": primal_calldata}
+            if self.d_in is not None:
+                if data.grad_in is None:
+                    raise ValueError("Missing required input gradients")
+                result["d_in"] = self.d_in.create_calldata(context, binding, data.grad_in)
+            if self.d_out is not None:
+                if data.grad_out is None:
+                    raise ValueError("Missing tensor to hold output gradients")
+                result["d_out"] = self.d_out.create_calldata(context, binding, data.grad_out)
+
+            if (
+                context.call_mode != CallMode.prim
+                and data.grad_in is not None
+                and data.grad_in is data.grad_out
+            ):
+                if binding.access[1] == AccessType.readwrite:
+                    raise ValueError(
+                        "inout parameter gradients need separate buffers for inputs and outputs (see Tensor.with_grads)"
+                    )
+        else:
+            primal_calldata = {
+                "buffer": primal.data_ptr(),
+                "layout": {"offset": 0, "strides": strides},
+                "_shape": shape,
+            }
+
+            if not self.d_in and not self.d_out:
+                return primal_calldata
+
+            result = {"primal": primal_calldata}
+            if self.d_in is not None:
+                if data.grad_in is None:
+                    raise ValueError("Missing required input gradients")
+                result["d_in"] = self.d_in.create_calldata(context, binding, data.grad_in)
+            if self.d_out is not None:
+                if data.grad_out is None:
+                    raise ValueError("Missing tensor to hold output gradients")
+                result["d_out"] = self.d_out.create_calldata(context, binding, data.grad_out)
+
+        return result
+
+    def read_calldata(
+        self,
+        context: CallContext,
+        binding: "BoundVariableRuntime",
+        data: TensorRef,
+        result: Any,
+    ):
+        if context.device.info.type != DeviceType.cuda:
+            assert data.tensor is not None
+            assert data.interop_buffer is not None
+            primal = cast(torch.Tensor, data.tensor)  # type: ignore
+
+            shape = tuple(primal.shape)
+            strides = primal.stride()
+            data_type = _torch_to_data_type[self.torch_dtype]
+            interop_tensor = cast(
+                torch.Tensor,
+                data.interop_buffer.to_torch(type=data_type, shape=shape, strides=strides),
+            )
+
+            primal.untyped_storage().copy_(interop_tensor.untyped_storage())
```

Comment:
This will fail if the size of the storage != size of the tensor. In `create_calldata`, you create an interop buffer of size `primal.numel() * primal.element_size()`, but the size of `primal`'s storage can be different.

For example:
```
a = torch.empty((64, 64))
b = a[0]
print(b.numel() * b.element_size()) # 256
print(b.untyped_storage().nbytes()) # 16384
```